File: old_src/tokenizer.lua
local function tokenizeCoroutine(str)
	local i = 1
	for j = 1, #str do
		local c = str:sub(j, j)
		if not c:match("%a") then
			if i < j then
				coroutine.yield {
					value = str:sub(i, j - 1),
					loc = {i, j - 1}
				}
			end
			if not c:match("%s") then
				coroutine.yield {
					value = c,
					loc = {j, j}
				}
			end
			i = j + 1
		end
	end
	if i <= #str then
		coroutine.yield {
			value = str:sub(i),
			loc = {i, #str}
		}
	end
end

local function tokenize(str)
	local i = str:find("[^%a%s():]")
	if i then
		error {
			msg = "Invalid character " .. str:sub(i, i),
			loc = {i, i}
		}
	end
	return coroutine.wrap(function() tokenizeCoroutine(str) end)
end

return tokenize

File: src/interpreter/tokenizer.ts
export enum TokenType {
    TEXT,
    SYMBOL,
};

type Token = { value: string, loc: [number, number], type: TokenType };

function* tokenizeImpl(str: string): Generator<Token, void, void> {
    let cur = [];
    let j = 0;
    for (let c of str) {
        j++;
        const is_letter = string.match(c, "%a") != null;
        const is_space = string.match(c, "%s");
        if (is_letter) {
            cur.push(c);
        } else {
            if (cur.length > 0) {
                yield {
                    value: cur.join(""),
                    loc: [j - cur.length, j - 1],
                    type: TokenType.TEXT
                };
                cur = [];
            }
            if (!is_space) {
                yield {
                    value: c,
                    loc: [j - 1, j - 1],
                    type: TokenType.SYMBOL
                }
            }
        }
    }
}

export function* tokenize(str: string): Generator<Token, void, void> {
    const result = string.find(str, "[^%a%s():]");
    // Can't to everything in one assignment because of a bug in tstl
    const [i, ...rest] = result;
    if (i != null) {
        throw { "msg": `Invalid character ${str.substr(i, i)}`, "loc": [i, i] };
    }
    yield* tokenizeImpl(str);
}